File: composeApp/src/commonMain/kotlin/com/soooool/matedash/data/api/TeslaMateApiClient.kt
package com.soooool.matedash.data.api

import com.soooool.matedash.data.model.ApiConfig
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.http.HttpHeaders
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

class TeslaMateApiClient {
    private val httpClient = HttpClient {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
            })
        }
    }

    suspend fun getCarStatus(config: ApiConfig): CarStatusDto {
        val response: ApiResponse = httpClient.get("${config.baseUrl}/api/v1/cars/${config.carId}/status") {
            if (config.apiToken.isNotBlank()) header(HttpHeaders.Authorization, "Bearer ${config.apiToken}")
        }.body()
        return response.data?.status ?: throw Exception("응답 데이터 없음")
    }

    suspend fun getCars(config: ApiConfig): List<CarListItemDto> {
        return try {
            httpClient.get("${config.baseUrl}/api/v1/cars") {
                if (config.apiToken.isNotBlank()) header(HttpHeaders.Authorization, "Bearer ${config.apiToken}")
            }.body()
        } catch (_: Exception) {
            emptyList()
        }
    }

    suspend fun getDrives(config: ApiConfig, limit: Int = 20): List<DriveDto> {
        val response: DrivesResponse = httpClient.get(
            "${config.baseUrl}/api/v1/cars/${config.carId}/drives?limit=$limit"
        ) {
            if (config.apiToken.isNotBlank()) header(HttpHeaders.Authorization, "Bearer ${config.apiToken}")
        }.body()
        return response.data?.drives ?: emptyList()
    }

    suspend fun getCharges(config: ApiConfig, limit: Int = 20): List<ChargeDto> {
        val response: ChargesResponse = httpClient.get(
            "${config.baseUrl}/api/v1/cars/${config.carId}/charges?limit=$limit"
        ) {
            if (config.apiToken.isNotBlank()) header(HttpHeaders.Authorization, "Bearer ${config.apiToken}")
        }.body()
        return response.data?.charges ?: emptyList()
    }

    fun close() = httpClient.close()
}

@kotlinx.serialization.Serializable
data class CarListItemDto(
    val id: Int? = null,
    val name: String? = null,
    val model: String? = null,
)

File: iosApp/MateDashWidgets/Networking/TeslaMateAPIClient.swift
import Foundation

struct TeslaMateAPIClient {
    let baseURL: String
    let apiToken: String
    let carId: Int

    init(config: ApiWidgetConfig) {
        self.baseURL = config.baseURL
        self.apiToken = config.apiToken
        self.carId = config.carId
    }

    private var decoder: JSONDecoder {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }

    private func request(path: String) -> URLRequest {
        let url = URL(string: "\(baseURL)\(path)")!
        var req = URLRequest(url: url, timeoutInterval: 15)
        if !apiToken.isEmpty {
            req.setValue("Bearer \(apiToken)", forHTTPHeaderField: "Authorization")
        }
        return req
    }

    func fetchCharges(limit: Int = 100) async throws -> [ChargeDTO] {
        let req = request(path: "/api/v1/cars/\(carId)/charges?limit=\(limit)")
        let (data, _) = try await URLSession.shared.data(for: req)
        let response = try decoder.decode(ChargesResponse.self, from: data)
        return response.data?.charges ?? []
    }

    func fetchCarStatus() async throws -> CarStatusDTO {
        let req = request(path: "/api/v1/cars/\(carId)/status")
        let (data, _) = try await URLSession.shared.data(for: req)
        let response = try decoder.decode(ApiStatusResponse.self, from: data)
        guard let status = response.data?.status else {
            throw URLError(.cannotParseResponse)
        }
        return status
    }

    func fetchDrives(limit: Int = 20) async throws -> [DriveDTO] {
        let req = request(path: "/api/v1/cars/\(carId)/drives?limit=\(limit)")
        let (data, _) = try await URLSession.shared.data(for: req)
        let response = try decoder.decode(DrivesResponse.self, from: data)
        return response.data?.drives ?? []
    }
}
